src/clims/models/extensible.py

```diff
@@ -1,10 +1,44 @@
 from __future__ import absolute_import, print_function
 
 from django.db import models
-from sentry.db.models import (Model, sane_repr)
+from sentry.db.models import (Model, sane_repr, FlexibleForeignKey)
 import json
 
 
+class ExtensibleModel(Model):
+    """
+    Base class for extensible models
+    """
+    __core__ = True
+
+    extensible_type = FlexibleForeignKey('clims.ExtensibleType')
+
+    class Meta:
+        abstract = True
+
+
+class ExtensibleVersion(Model):
+    """
+    Represents a particular version of an Extensible
+
+    Parent/Child relationship is always between a particular version of an Extensible.
+    """
+    __core__ = True
+
+    # This field is only populated if a user renames the extensible. In this case we record
+    # the name change here in the version model. This makes handling unique constraints simpler.
+    previous_name = models.TextField(null=True)
```

Comment:
Ahh. I was thinking of changing this to name and populate it always with the current name. It will be clearer and this approach was used as I was probably in a micro optimizing mode/daze. We can look into that later.

But this is when you want to rename the instance. So the idea is that the `Extensible` model always has the current name of the thing in question (e.g. `sample-1`), but the `ExtensibleVersion` has the whole history (either using this approach or the aforementioned approach of always saving the current name).